File: dotnet/synthings.library.tests/Libraries.fs
module synthings.library.tests.Libraries

open Xunit
open synthings.core

[<Fact>]
let ``Find a topic in the aggregate library`` () =
    //assumes a reference to synthings.scalar
    let library = aggregateLibrary.build()
    let topics = library.listTopics()
    let waveTopics = List.filter (fun (topic : TopicDescriptor) -> topic.DisplayName.Contains "Wave") topics
    Assert.True(waveTopics.Length = 1)
    
[<Fact>]
let ``Create a machine from a behavior in the aggregate library`` () =
    //assumes a reference to synthings.scalar
    let library = aggregateLibrary.build()
    let topics = library.listTopics()
    let waveTopic = List.find (fun (topic : TopicDescriptor) -> topic.DisplayName.Contains "Wave") topics
    let waveBehaviors = library.listBehaviors waveTopic
    let sineBehavior = List.find (fun (behaviorDescriptor : BehaviorDescriptor) -> behaviorDescriptor.DisplayName.Contains "Sine") waveBehaviors
    let sineMachine = library.createMachine sineBehavior
    Assert.True(sineMachine.Name.Contains "Sine")

File: dotnet/synthings.csharp.tests/Libraries.cs
namespace synthings.csharp.tests
{
    using Xunit;
    using System.Linq;
    using csharp;

    public class Libraries
    {
        [Fact]
        public void Load_the_aggregate_library()
        {
            var library = Library.Load();
            Assert.Equal("Aggregate", library.Name);
        }

        [Fact]
        public void Find_a_Behavior()
        {
            var library = Library.Load();
            var topics = library.listTopics();
            var waveTopic = topics.First(topic => topic.DisplayName.Contains("Wave"));
            var behaviors = library.listBehaviors(waveTopic);
            var sineWaveBehavior = behaviors.First(behavior => behavior.DisplayName.Contains("Sine"));
            Assert.Equal("Sine Wave", sineWaveBehavior.DisplayName);
        }

        [Fact]
        public void Create_a_machine()
        {
            var library = Library.Load();
            var topics = library.listTopics();
            var envelopeTopic = topics.First(topic => topic.DisplayName.Contains("Envelope"));
            var behaviors = library.listBehaviors(envelopeTopic);
            var linearDecayBehavior = behaviors.First(behavior => behavior.DisplayName.Contains("Linear"));
            var linearDecayMachine = library.createMachine(linearDecayBehavior);
            Assert.Equal("Linear Decay", linearDecayMachine.Name);
        }
    }
}